src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -186,10 +186,13 @@ public String url() {
      * Flag for whether or not the build has completed somehow.
      * This was previously a transient field, so we may need to recompute in {@link #onLoad} based on {@link FlowExecution#isComplete}.
      */
-    Boolean completed;  // Non-private for testing
+    volatile Boolean completed;  // Non-private for testing
 
-    /** Protects access to {@link #completed} etc. */
-    private transient Object metadataGuard = new Object();
+    /**
+     * Protects access to {@link #completed} etc.
+     * @see #getMetadataGuard
+     */
+    private transient volatile Object metadataGuard = new Object();
```

Comment:
Needs to be volatile now that `getMetadataGuard` does not lock the `WorkflowRun` for its entire body.